File: ios/Photo/UI/LayerOverlayView.swift
import UIKit

/// Reusable native move/scale/rotate surface shared by every overlay type.
final class LayerOverlayView: UIView {
  private enum Gesture { case none, drag, handle, multi }
  private var imageBounds: CGRect = .zero
  var minScale: CGFloat = 0.2
  var maxScale: CGFloat = 8
  var layers: [PhotoLayer] = [] { didSet { setNeedsDisplay() } }
  var selectedLayerID: String? { didSet { setNeedsDisplay() } }
  var onLayerTapped: ((String?) -> Void)?
  var onLayerDoubleTapped: ((String) -> Void)?
  var onLayerTransformChanged: ((_ id: String, _ x: CGFloat, _ y: CGFloat, _ scale: CGFloat, _ rotationDegrees: CGFloat) -> Void)?
  var onLayerTransformEnded: ((String) -> Void)?

  private var gesture: Gesture = .none
  private var gestureID: String?
  private var startLayer: PhotoLayer?
  private var primaryTouch: UITouch?
  private var secondaryTouch: UITouch?
  private var startPoint: CGPoint = .zero
  private var startDistance: CGFloat = 1
  private var startAngle: CGFloat = 0
  private var startMidpoint: CGPoint = .zero
  private let handleRadius: CGFloat = 11
  private let handleTouchRadius: CGFloat = 30

  override init(frame: CGRect) {
    super.init(frame: frame)
    backgroundColor = .clear
    isOpaque = false
    isMultipleTouchEnabled = true
  }
  required init?(coder: NSCoder) { nil }

  func setImageBounds(_ bounds: CGRect) { imageBounds = bounds; setNeedsDisplay() }

  override func draw(_ rect: CGRect) {
    guard let layer = selectedLayer(), let context = UIGraphicsGetCurrentContext() else { return }
    let center = center(of: layer); let half = selectionHalfExtents(layer)
    context.saveGState()
    context.translateBy(x: center.x, y: center.y)
    context.rotate(by: layer.rotationDegrees * .pi / 180)
    context.setStrokeColor(UIColor.white.cgColor)
    context.setLineWidth(2)
    context.setLineDash(phase: 0, lengths: [8, 5])
    context.stroke(CGRect(x: -half.width, y: -half.height, width: half.width * 2, height: half.height * 2))
    context.setLineDash(phase: 0, lengths: [])
    context.setFillColor(UIColor.white.cgColor)
    context.fillEllipse(in: CGRect(x: half.width - handleRadius, y: half.height - handleRadius, width: handleRadius * 2, height: handleRadius * 2))
    context.setStrokeColor(UIColor(red: 87/255, green: 55/255, blue: 245/255, alpha: 1).cgColor)
    context.setLineWidth(2)
    let glyph = handleRadius * 0.52
    context.move(to: CGPoint(x: half.width - glyph, y: half.height + glyph)); context.addLine(to: CGPoint(x: half.width + glyph, y: half.height - glyph)); context.strokePath()
    context.restoreGState()
  }

  override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
    guard imageBounds.width > 0 else { return }
    for touch in touches {
      if primaryTouch == nil { primaryTouch = touch }
      else if secondaryTouch == nil { secondaryTouch = touch }
    }
    guard let primaryTouch else { return }
    let point = primaryTouch.location(in: self)
    if let secondaryTouch, let current = currentGestureLayer() {
      startLayer = current
      let metrics = multiMetrics(primaryTouch, secondaryTouch)
      startDistance = max(1, metrics.distance); startAngle = metrics.angle; startMidpoint = metrics.midpoint
      gesture = .multi
      return
    }
    let selected = selectedLayer()
    let handleHit = selected.flatMap { !$0.locked && isHandleHit($0, point) ? $0 : nil }
    let hit = handleHit ?? hitTest(point)
    onLayerTapped?(hit?.id)
    guard let hit, !hit.locked else { resetGesture(); return }
    gestureID = hit.id; startLayer = hit; startPoint = point
    if handleHit != nil {
      gesture = .handle
      let center = center(of: hit)
      startDistance = max(1, distance(center, point)); startAngle = angle(center, point)
    } else { gesture = .drag }
  }

  override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
    guard let initial = startLayer, let primaryTouch else { return }
    let point = primaryTouch.location(in: self)
    switch gesture {
    case .drag:
      let x = clamp(initial.x + (point.x - startPoint.x) / imageBounds.width, 0, 1)
      let y = clamp(initial.y + (point.y - startPoint.y) / imageBounds.height, 0, 1)
      emit(initial, x: x, y: y, scale: initial.scale, rotation: initial.rotationDegrees)
    case .handle:
      let center = center(of: initial)
      let scale = clamp(initial.scale * distance(center, point) / startDistance, minScale, maxScale)
      let rotation = normalizeDegrees(initial.rotationDegrees + angleDelta(from: startAngle, to: angle(center, point)))
      emit(initial, x: initial.x, y: initial.y, scale: scale, rotation: rotation)
    case .multi:
      guard let secondaryTouch else { return }
      let metrics = multiMetrics(primaryTouch, secondaryTouch)
      let scale = clamp(initial.scale * metrics.distance / startDistance, minScale, maxScale)
      let rotation = normalizeDegrees(initial.rotationDegrees + angleDelta(from: startAngle, to: metrics.angle))
      let x = clamp(initial.x + (metrics.midpoint.x - startMidpoint.x) / imageBounds.width, 0, 1)
      let y = clamp(initial.y + (metrics.midpoint.y - startMidpoint.y) / imageBounds.height, 0, 1)
      emit(initial, x: x, y: y, scale: scale, rotation: rotation)
    case .none: break
    }
  }

  override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
    if let touch = touches.first, touch.tapCount >= 2, let id = gestureID { onLayerDoubleTapped?(id) }
    endTouches(touches)
  }
  override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) { endTouches(touches) }

  private func endTouches(_ touches: Set<UITouch>) {
    let endedPrimary = primaryTouch.map { touches.contains($0) } ?? false
    let endedSecondary = secondaryTouch.map { touches.contains($0) } ?? false
    if endedPrimary || endedSecondary { finishGesture() }
  }
  private func emit(_ layer: PhotoLayer, x: CGFloat, y: CGFloat, scale: CGFloat, rotation: CGFloat) {
    onLayerTransformChanged?(layer.id, x, y, scale, rotation)
  }
  private func finishGesture() {
    if let id = gestureID { onLayerTransformEnded?(id) }
    resetGesture()
  }
  private func resetGesture() {
    gesture = .none; gestureID = nil; startLayer = nil; primaryTouch = nil; secondaryTouch = nil
  }
  private func selectedLayer() -> PhotoLayer? { layers.first { $0.id == selectedLayerID && $0.visible } }
  private func currentGestureLayer() -> PhotoLayer? { layers.first { $0.id == gestureID && $0.visible && !$0.locked } }
  private func center(of layer: PhotoLayer) -> CGPoint { CGPoint(x: imageBounds.minX + layer.x * imageBounds.width, y: imageBounds.minY + layer.y * imageBounds.height) }
  private func selectionHalfExtents(_ layer: PhotoLayer) -> CGSize {
    let shortSide = min(imageBounds.width, imageBounds.height)
    let base: CGSize
    switch layer.type {
    case .text:
      let pointSize = layer.fontSize / 48 * shortSide * 0.06
      let font = layer.fontFamily.flatMap { UIFont(name: $0, size: pointSize) } ?? UIFont.systemFont(ofSize: pointSize)
      let lines = layer.text.isEmpty ? [" "] : layer.text.components(separatedBy: "\n")
      let width = lines.map { ($0 as NSString).size(withAttributes: [.font: font]).width }.max() ?? pointSize
      base = CGSize(width: width / 2 + 8, height: font.lineHeight * CGFloat(lines.count) / 2 + 6)
    case .sticker:
      // Mirrors `PhotoLayerRenderer.drawSticker`'s `shortSide * 0.18` box (half-extent = 0.09).
      let half = shortSide * 0.09
      base = CGSize(width: half, height: half)
    case .overlay:
      // Mirrors `PhotoLayerRenderer.drawOverlay`'s box: full canvas width, height from aspect ratio.
      let aspectRatio = layer.overlayAspectRatio.flatMap { $0 > 0 ? $0 : nil } ?? 1
      base = CGSize(width: imageBounds.width / 2, height: imageBounds.width / (2 * aspectRatio))
    case .drawing:
      let maxX = layer.drawPoints.map { abs($0.0) }.max() ?? 0.08
      let maxY = layer.drawPoints.map { abs($0.1) }.max() ?? 0.08
      base = CGSize(width: maxX * imageBounds.width + handleRadius, height: maxY * imageBounds.height + handleRadius)
    }
    return CGSize(width: max(handleRadius, base.width * layer.scale), height: max(handleRadius, base.height * layer.scale))
  }
  private func handlePoint(_ layer: PhotoLayer) -> CGPoint {
    let center = center(of: layer); let half = selectionHalfExtents(layer); let radians = layer.rotationDegrees * .pi / 180
    return CGPoint(x: center.x + half.width * cos(radians) - half.height * sin(radians), y: center.y + half.width * sin(radians) + half.height * cos(radians))
  }
  private func isHandleHit(_ layer: PhotoLayer, _ point: CGPoint) -> Bool { distance(handlePoint(layer), point) <= handleTouchRadius }
  private func hitTest(_ point: CGPoint) -> PhotoLayer? {
    for layer in layers.reversed() where layer.visible {
      let center = center(of: layer); let radians = -layer.rotationDegrees * .pi / 180
      let dx = point.x - center.x; let dy = point.y - center.y
      let localX = dx * cos(radians) - dy * sin(radians); let localY = dx * sin(radians) + dy * cos(radians)
      let half = selectionHalfExtents(layer)
      if abs(localX) <= half.width && abs(localY) <= half.height { return layer }
    }
    return nil
  }
  private func multiMetrics(_ first: UITouch, _ second: UITouch) -> (midpoint: CGPoint, distance: CGFloat, angle: CGFloat) {
    let a = first.location(in: self); let b = second.location(in: self)
    return (CGPoint(x: (a.x + b.x) / 2, y: (a.y + b.y) / 2), distance(a, b), angle(a, b))
  }
  private func distance(_ a: CGPoint, _ b: CGPoint) -> CGFloat { hypot(b.x - a.x, b.y - a.y) }
  private func angle(_ a: CGPoint, _ b: CGPoint) -> CGFloat { atan2(b.y - a.y, b.x - a.x) * 180 / .pi }
  private func angleDelta(from: CGFloat, to: CGFloat) -> CGFloat { (to - from + 540).truncatingRemainder(dividingBy: 360) - 180 }
  private func normalizeDegrees(_ value: CGFloat) -> CGFloat { (value.truncatingRemainder(dividingBy: 360) + 360).truncatingRemainder(dividingBy: 360) }
  private func clamp(_ value: CGFloat, _ minimum: CGFloat, _ maximum: CGFloat) -> CGFloat { min(max(value, minimum), maximum) }
}

File: android/src/main/java/com/photovideoeditor/photo/ui/LayerOverlayView.kt
package com.photovideoeditor.photo.ui

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.view.MotionEvent
import android.view.View
import android.text.TextPaint
import com.photovideoeditor.R
import com.photovideoeditor.photo.render.OverlayGeometry
import com.photovideoeditor.photo.render.PhotoLayer
import com.photovideoeditor.photo.render.LayerType
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.hypot
import kotlin.math.min
import kotlin.math.sin

/** Reusable native move/scale/rotate surface shared by every overlay type. */
class LayerOverlayView(context: Context) : View(context) {
  private enum class Gesture { NONE, DRAG, HANDLE, MULTI }
  private var imageBounds = RectF()
  var minScale = 0.2f
  var maxScale = 8f
  var layers: List<PhotoLayer> = emptyList()
    set(value) { field = value; invalidate() }
  var selectedLayerId: String? = null
    set(value) { field = value; invalidate() }
  var onLayerDelete: ((String) -> Unit)? = null
  var onLayerTapped: ((String?) -> Unit)? = null
  var onLayerDoubleTapped: ((String) -> Unit)? = null
  var onLayerTransformChanged: ((id: String, x: Float, y: Float, scale: Float, rotationDegrees: Float) -> Unit)? = null
  var onLayerTransformEnded: ((id: String) -> Unit)? = null

  private var gesture = Gesture.NONE
  private var gestureId: String? = null
  private var startLayer: PhotoLayer? = null
  private var startPointX = 0f
  private var startPointY = 0f
  private var startDistance = 1f
  private var startAngle = 0f
  private var startMidX = 0f
  private var startMidY = 0f
  private var movedDuringGesture = false
  private var lastTapTimeMs = 0L
  private var lastTapLayerId: String? = null
  private val density = resources.displayMetrics.density
  private val handleRadius = 11f * density
  private val handleTouchRadius = 30f * density
  private val selectionPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    color = Color.WHITE; style = Paint.Style.STROKE; strokeWidth = 1f * density
  }
  private val handlePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = Color.WHITE }


  fun setImageBounds(bounds: RectF) { imageBounds = RectF(bounds); invalidate() }

  override fun onDraw(canvas: Canvas) {
    super.onDraw(canvas)
    val selected = selectedLayer() ?: return
    val center = centerOf(selected)
    val half = selectionHalfExtents(selected)
    canvas.save()
    canvas.translate(center.first, center.second)
    canvas.rotate(selected.rotationDegrees)
    canvas.drawRect(-half.first, -half.second, half.first, half.second, selectionPaint)
    for ((x, y) in listOf(Pair(-half.first, -half.second), Pair(half.first, -half.second), Pair(-half.first, half.second), Pair(half.first, half.second))) {
      canvas.drawCircle(x, y, 4f * density, handlePaint)
    }
    drawControl(canvas, -half.first, -half.second, R.drawable.ic_close, Color.rgb(255, 100, 115))
    drawControl(canvas, half.first, half.second, R.drawable.ic_fullscreen, Color.rgb(167, 139, 250))
    canvas.restore()
  }

  private fun drawControl(canvas: Canvas, x: Float, y: Float, icon: Int, tint: Int) {
    canvas.drawCircle(x, y, handleRadius, handlePaint)
    context.getDrawable(icon)?.mutate()?.let { drawable ->
      drawable.setTint(tint)
      val size = (handleRadius * 0.7f).toInt()
      drawable.setBounds(x.toInt() - size, y.toInt() - size, x.toInt() + size, y.toInt() + size)
      drawable.draw(canvas)
    }
  }

  override fun onTouchEvent(event: MotionEvent): Boolean {
    if (imageBounds.width() <= 0 || imageBounds.height() <= 0) return false
    when (event.actionMasked) {
      MotionEvent.ACTION_DOWN -> {
        val selected = selectedLayer()
        if (selected != null && !selected.locked) {
          val handle = handlePoint(selected)
          val center = centerOf(selected)
          if (hypot(event.x - (2 * center.first - handle.first), event.y - (2 * center.second - handle.second)) <= handleTouchRadius) {
            onLayerDelete?.invoke(selected.id); resetGesture(); return true
          }
        }
        val handleHit = selected?.takeIf { !it.locked && isHandleHit(it, event.x, event.y) }
        val hit = handleHit ?: hitTest(event.x, event.y)
        onLayerTapped?.invoke(hit?.id)
        if (hit == null || hit.locked) { resetGesture(); return false }
        gestureId = hit.id
        movedDuringGesture = false
        startLayer = hit.copy()
        startPointX = event.x; startPointY = event.y
        if (handleHit != null) {
          gesture = Gesture.HANDLE
          val center = centerOf(hit)
          startDistance = hypot(event.x - center.first, event.y - center.second).coerceAtLeast(1f)
          startAngle = angle(center.first, center.second, event.x, event.y)
        } else gesture = Gesture.DRAG
        parent?.requestDisallowInterceptTouchEvent(true)
        return true
      }
      MotionEvent.ACTION_POINTER_DOWN -> {
        val layer = currentGestureLayer() ?: return false
        if (event.pointerCount >= 2) {
          startLayer = layer.copy()
          val metrics = multiMetrics(event)
          startDistance = metrics.distance.coerceAtLeast(1f); startAngle = metrics.angle
          startMidX = metrics.midX; startMidY = metrics.midY
          movedDuringGesture = true
          gesture = Gesture.MULTI
          return true
        }
      }
      MotionEvent.ACTION_MOVE -> {
        val initial = startLayer ?: return false
        if (hypot(event.x - startPointX, event.y - startPointY) > handleRadius) movedDuringGesture = true
        when (gesture) {
          Gesture.DRAG -> {
            val x = (initial.x + (event.x - startPointX) / imageBounds.width()).coerceIn(0f, 1f)
            val y = (initial.y + (event.y - startPointY) / imageBounds.height()).coerceIn(0f, 1f)
            emit(initial, x, y, initial.scale, initial.rotationDegrees)
          }
          Gesture.HANDLE -> {
            val center = centerOf(initial)
            val scale = (initial.scale * hypot(event.x - center.first, event.y - center.second) / startDistance).coerceIn(minScale, maxScale)
            val rotation = normalizeDegrees(initial.rotationDegrees + angleDelta(startAngle, angle(center.first, center.second, event.x, event.y)))
            emit(initial, initial.x, initial.y, scale, rotation)
          }
          Gesture.MULTI -> if (event.pointerCount >= 2) {
            val metrics = multiMetrics(event)
            val scale = (initial.scale * metrics.distance / startDistance).coerceIn(minScale, maxScale)
            val rotation = normalizeDegrees(initial.rotationDegrees + angleDelta(startAngle, metrics.angle))
            val x = (initial.x + (metrics.midX - startMidX) / imageBounds.width()).coerceIn(0f, 1f)
            val y = (initial.y + (metrics.midY - startMidY) / imageBounds.height()).coerceIn(0f, 1f)
            emit(initial, x, y, scale, rotation)
          }
          Gesture.NONE -> return false
        }
        return true
      }
      MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
        val tappedId = gestureId
        if (event.actionMasked == MotionEvent.ACTION_UP && !movedDuringGesture && tappedId != null) {
          val now = android.os.SystemClock.uptimeMillis()
          if (lastTapLayerId == tappedId && now - lastTapTimeMs <= 350L) {
            onLayerDoubleTapped?.invoke(tappedId); lastTapLayerId = null; lastTapTimeMs = 0L
          } else { lastTapLayerId = tappedId; lastTapTimeMs = now }
        }
        finishGesture(); return true
      }
      MotionEvent.ACTION_POINTER_UP -> {
        // Rebase to the remaining finger so lifting a pinch finger does not jump/freeze the layer.
        val remaining = if (event.actionIndex == 0) 1 else 0
        if (remaining < event.pointerCount) {
          startLayer = currentGestureLayer()?.copy()
          startPointX = event.getX(remaining); startPointY = event.getY(remaining)
          gesture = Gesture.DRAG
        }
        movedDuringGesture = true
        return true
      }
    }
    return false
  }

  private fun emit(layer: PhotoLayer, x: Float, y: Float, scale: Float, rotation: Float) =
    onLayerTransformChanged?.invoke(layer.id, x, y, scale, rotation)

  private fun finishGesture() {
    gestureId?.let { onLayerTransformEnded?.invoke(it) }
    resetGesture()
    parent?.requestDisallowInterceptTouchEvent(false)
  }

  private fun resetGesture() { gesture = Gesture.NONE; gestureId = null; startLayer = null }
  private fun selectedLayer() = layers.firstOrNull { it.id == selectedLayerId && it.visible }
  private fun currentGestureLayer() = layers.firstOrNull { it.id == gestureId && it.visible && !it.locked }
  private fun centerOf(layer: PhotoLayer) = Pair(imageBounds.left + layer.x * imageBounds.width(), imageBounds.top + layer.y * imageBounds.height())
  private fun selectionHalfExtents(layer: PhotoLayer): Pair<Float, Float> {
    val shortSide = min(imageBounds.width(), imageBounds.height())
    val base = when (layer.type) {
      LayerType.TEXT -> {
        val paint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
          textSize = layer.fontSize / 48f * shortSide * 0.06f
          layer.fontFamily?.let { typeface = android.graphics.Typeface.create(it, android.graphics.Typeface.NORMAL) }
        }
        val lines = layer.text.ifBlank { " " }.split("\n")
        Pair((lines.maxOfOrNull { paint.measureText(it) } ?: paint.textSize) / 2f + 8f * density, paint.fontSpacing * lines.size / 2f + 6f * density)
      }
      LayerType.STICKER -> {
        val (width, height) = OverlayGeometry.stickerSize(shortSide, layer.overlayAspectRatio ?: 1f)
        Pair(width / 2f + 4f * density, height / 2f + 4f * density)
      }
      LayerType.OVERLAY -> {
        val aspect = layer.overlayAspectRatio ?: 1f
        val baseSize = shortSide * 0.6f
        if (aspect >= 1f) Pair(baseSize / 2f, (baseSize / aspect) / 2f) else Pair((baseSize * aspect) / 2f, baseSize / 2f)
      }
      LayerType.DRAWING -> {
        val maxX = layer.drawPoints.maxOfOrNull { kotlin.math.abs(it.first) } ?: 0.08f
        val maxY = layer.drawPoints.maxOfOrNull { kotlin.math.abs(it.second) } ?: 0.08f
        Pair(maxX * imageBounds.width() + handleRadius, maxY * imageBounds.height() + handleRadius)
      }
    }
    return Pair((base.first * layer.scale).coerceAtLeast(24f * density), (base.second * layer.scale).coerceAtLeast(24f * density))
  }

  private fun handlePoint(layer: PhotoLayer): Pair<Float, Float> {
    val center = centerOf(layer); val half = selectionHalfExtents(layer)
    val radians = Math.toRadians(layer.rotationDegrees.toDouble())
    return Pair(center.first + (half.first * cos(radians) - half.second * sin(radians)).toFloat(), center.second + (half.first * sin(radians) + half.second * cos(radians)).toFloat())
  }

  private fun isHandleHit(layer: PhotoLayer, x: Float, y: Float): Boolean {
    val handle = handlePoint(layer)
    return hypot(x - handle.first, y - handle.second) <= handleTouchRadius
  }

  private fun hitTest(x: Float, y: Float): PhotoLayer? = layers.asReversed().firstOrNull { layer ->
    if (!layer.visible) return@firstOrNull false
    val center = centerOf(layer); val radians = Math.toRadians((-layer.rotationDegrees).toDouble())
    val dx = x - center.first; val dy = y - center.second
    val localX = dx * cos(radians).toFloat() - dy * sin(radians).toFloat()
    val localY = dx * sin(radians).toFloat() + dy * cos(radians).toFloat()
    val half = selectionHalfExtents(layer)
    localX in -half.first..half.first && localY in -half.second..half.second
  }

  private data class MultiMetrics(val midX: Float, val midY: Float, val distance: Float, val angle: Float)
  private fun multiMetrics(event: MotionEvent): MultiMetrics {
    val x0 = event.getX(0); val y0 = event.getY(0); val x1 = event.getX(1); val y1 = event.getY(1)
    return MultiMetrics((x0 + x1) / 2f, (y0 + y1) / 2f, hypot(x1 - x0, y1 - y0), angle(x0, y0, x1, y1))
  }
  private fun angle(x0: Float, y0: Float, x1: Float, y1: Float) = Math.toDegrees(atan2(y1 - y0, x1 - x0).toDouble()).toFloat()
  private fun angleDelta(from: Float, to: Float): Float = ((to - from + 540f) % 360f) - 180f
  private fun normalizeDegrees(value: Float): Float = ((value % 360f) + 360f) % 360f
}
